pgrx/src/pg_catalog.rs

```diff
@@ -0,0 +1,1212 @@
+                        _ => panic!("unrecognized value: `{}`", value as u8 as char),
+                    }
+                }
+                pub fn as_c_char(self) -> c_char {
+                    self as u8 as c_char
+                }
+            }
+
+            impl pgrx::datum::FromDatum for [<$table:camel $column:camel>] {
+                unsafe fn from_polymorphic_datum(
+                    datum: pgrx::pg_sys::Datum,
+                    is_null: bool,
+                    _: Oid,
+                ) -> Option<Self>
+                where
+                    Self: Sized,
+                {
+                    if is_null {
+                        None
+                    } else {
+                        Some(Self::new(datum.value() as _))
+                    }
+                }
+            }
+
+            impl pgrx::datum::IntoDatum for [<$table:camel $column:camel>] {
+                fn into_datum(self) -> std::option::Option<pgrx::pg_sys::Datum> {
+                    Some(pgrx::pg_sys::Datum::from(self as i8))
+                }
+                fn type_oid() -> pgrx_pg_sys::Oid {
+                    pgrx::pg_sys::CHAROID
+                }
+            }
+
+            unsafe impl pgrx::datum::UnboxDatum for [<$table:camel $column:camel>] {
+                type As<'src> = Self;
+                #[inline]
+                unsafe fn unbox<'src>(datum: pgrx::datum::Datum<'src>) -> Self::As<'src>
+                where
+                    Self: 'src,
+                {
+                    Self::new(datum.sans_lifetime().value() as c_char)
+                }
+            }
+
+            unsafe impl GetStruct<c_char> for [<$table:camel $column:camel>] {
+                unsafe fn get_struct(raw: *const c_char) -> Self {
+                    unsafe { Self::new(raw.read()) }
+                }
+            }
+        }
+    };
+}
+
+macro_rules! define_column {
+    {
+        $(#[$column_meta:meta])* column ($table:ident, $column:ident, $column_type:ty, $(#[$character_meta:meta])* character { $($(#[$variant_meta:meta])* ($variant:ident, $variant_value:literal))* } $($x:tt)*)
+    } => {
+        paste::paste! {
+            define_character! { $(#[$character_meta])* character ($table, $column) { $($(#[$variant_meta])* ($variant, $variant_value))* } }
+            define_column! { $(#[$column_meta])* column ($table, $column, $column_type $($x)*) }
+        }
+    };
+    {
+        $(#[$column_meta:meta])* column ($table:ident, $column:ident, $column_type:ty, get_struct)
+    } => {
+        paste::paste! {
+            impl [<$table:camel>]<'_> {
+                $(#[$column_meta])*
+                pub fn $column(&self) -> $column_type {
+                    unsafe {
+                        let start = self.inner.t_data.cast::<u8>();
+                        let offset = (*self.inner.t_data).t_hoff as usize;
+                        let p = start.add(offset).cast::<pgrx::pg_sys::[<FormData_ $table>]>();
```

Comment:
Idiomatic Rust normally avoids type annotations, but I would prefer these macros enforce their expectations by declaring relevant types like we were back in the C Mines and tending JavaBean Farms, especially before casting the type away and feeding it into a pointer offset.
```suggestion
                        let offset: u8 = (*self.inner.t_data).t_hoff;
                        let p = start.add(offset as usize).cast::<pgrx::pg_sys::[<FormData_ $table>]>();
```

Reply:
Done.